mountpoint-s3-crt/src/s3/client.rs

```diff
@@ -545,11 +556,117 @@ impl Drop for MetaRequest {
+        // SAFETY: self.inner is a valid aws_s3_meta_request and aws_s3_meta_request_acquire
+        // increments the reference count for it (and always returns a copy of the input, which is non-null).
+        let inner = unsafe { NonNull::new_unchecked(aws_s3_meta_request_acquire(self.inner.as_ptr())) };
+        Self { inner }
+    }
+}
+
 // SAFETY: `aws_s3_meta_request` is thread-safe
 unsafe impl Send for MetaRequest {}
 // SAFETY: `aws_s3_meta_request` is thread safe
 unsafe impl Sync for MetaRequest {}
 
+/// Future returned by `MetaRequest::write()`. It will complete when the write completes,
+/// or cancel the meta-request if dropped.
+#[derive(Debug)]
+pub struct MetaRequestWrite<'a> {
+    /// Signals when the write operation completes
+    receiver: oneshot::Receiver<Result<(), Error>>,
+    /// The meta-request to cancel if this future is dropped
+    request: MetaRequest,
+    _slice: PhantomData<&'a u8>,
+}
+
+impl Drop for MetaRequestWrite<'_> {
+    fn drop(&mut self) {
+        if !self.receiver.is_terminated() {
+            // This future is being dropped before completion. Cancel the meta-request to ensure
+            // the client will not try to use the provided buffer.
+            self.request.cancel();
+        }
+    }
+}
+
+impl<'a> MetaRequestWrite<'a> {
+    fn new(meta_request: &MetaRequest, slice: &'a [u8], eof: bool) -> Self {
+        let request = meta_request.clone();
+
+        // SAFETY: `MetaRequestWrite` will ensure that `slice` is alive until the future completes or
+        // that the meta-request is canceled if the future is dropped, preventing further use of the
+        // `aws_byte_cursor`.
+        let data = unsafe { slice.as_aws_byte_cursor() };
+
+        // SAFETY: `aws_s3_meta_request_write` never returns NULL.
+        let future = unsafe { NonNull::new_unchecked(aws_s3_meta_request_write(request.inner.as_ptr(), data, eof)) };
+
+        let (sender, receiver) = oneshot::channel();
+        let user_data = Box::leak(Box::new(FutureVoidCallbackData { future, sender })) as *mut FutureVoidCallbackData
+            as *mut ::libc::c_void;
+
+        // SAFETY: `future.as_ptr()` is a valid `aws_future_void` and this is the only callback we are registering.
+        unsafe {
+            aws_future_void_register_callback(future.as_ptr(), Some(future_void_callback), user_data);
+        }
+
+        Self {
+            receiver,
+            request,
+            _slice: Default::default(),
+        }
+    }
+}
+
+impl Future for MetaRequestWrite<'_> {
+    type Output = Result<(), Error>;
+
+    fn poll(mut self: Pin<&mut Self>, cx: &mut std::task::Context<'_>) -> std::task::Poll<Self::Output> {
+        Pin::new(&mut self.receiver)
+            .poll(cx)
+            .map(|r| r.expect("receiver is never canceled"))
+    }
+}
+
+/// Safety: Don't call this function directly, only called by the CRT as a callback.
+unsafe extern "C" fn future_void_callback(user_data: *mut ::libc::c_void) {
```

Comment:
Is it possible for us to free the memory allocated for the written slice synchronously in this callback? For now it seems to be asynchronous, which may result in unnecessarily high memory usage (asynchronous free'ing does not adjust to the frequent writes). This is worsened by the fact that CRT copies our data and completes the future immediately: 
https://github.com/awslabs/aws-c-s3/blob/async-write-p3/source/s3_meta_request.c#L2260

Reply:
Not entirely sure I understand the question: the original slice is owned by the fuse request, which will release (or rather reuse) it after the write operation completes.